File: Geometry/Point.py
from decimal import Decimal, getcontext
from math import sqrt, acos

# Ajustar precisión
getcontext().prec = 30

class Point3d:
    def __init__(self, x=Decimal(0), y=Decimal(0), z=Decimal(0)):
        self.x = Decimal(x)
        self.y = Decimal(y)
        self.z = Decimal(z)

    # Operadores
    def __add__(self, other): # Suma de vectores
        return Point3d(self.x + other.x, self.y + other.y, self.z + other.z)

    def __sub__(self, other): # Resta de vectores
        return Point3d(self.x - other.x, self.y - other.y, self.z - other.z)

    def __mul__(self, t): #Mult por escalar (Aca toca por derecha)
        if isinstance(t, (int, float, Decimal)):
            return Point3d(self.x * Decimal(t), self.y * Decimal(t), self.z * Decimal(t))
        raise TypeError("Multiplicación no soportada")

    def __rmul__(self, t): 
        return self.__mul__(t)

    def __truediv__(self, t):
        return Point3d(self.x / Decimal(t), self.y / Decimal(t), self.z / Decimal(t))

    def __eq__(self, other):
        EPS = Decimal("1e-6")
        return abs(self.x - other.x) < EPS and abs(self.y - other.y) < EPS and abs(self.z - other.z) < EPS

    # Métodos estáticos
    @staticmethod
    def dot(a, b):
        return a.x * b.x + a.y * b.y + a.z * b.z

    @staticmethod
    def cross(a, b):
        return Point3d(
            a.y * b.z - a.z * b.y,
            a.z * b.x - a.x * b.z,
            a.x * b.y - a.y * b.x
        )

    @staticmethod
    def norm(a):
        return Point3d.dot(a, a)

    @staticmethod
    def abs(a):
        return Decimal(sqrt(float(Point3d.norm(a))))  # sqrt no soporta Decimal directamente

    @staticmethod
    def proj(a, b):
        return b * (Point3d.dot(a, b) / Point3d.abs(b))

    @staticmethod
    def angle(a, b):
        return Decimal(acos(float(Point3d.dot(a, b) / (Point3d.abs(a) * Point3d.abs(b)))))


t = Decimal("0.5")
OA = Point3d(3, 1, 4.5)
OB = Point3d(3, 4, 0)

print("OA+OB =", (OA+OB).x, (OA+OB).y, (OA+OB).z)
print("OA-OB =", (OA-OB).x, (OA-OB).y, (OA-OB).z)
print("tOB =", (OB*t).x, (OB*t).y, (OB*t).z)
print("(1/t)OB =", (OB/t).x, (OB/t).y, (OB/t).z)
cross = Point3d.cross(OA, OB)
print("OAxOB =", cross.x, cross.y, cross.z)
print("OA*OB =", Point3d.dot(OA, OB))
print("||OB|| =", Point3d.abs(OB))
proj = Point3d.proj(OA, OB)
print("Proyeccion de OA sobre OB =", proj.x, proj.y, proj.z)
print("OA = OB", OA == OB)

File: Geometry/Point.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define rall(x) x.rbegin(), x.rend()
#define MOD 998244353
// #define MOD pow(10, 9)+7;
// #define MOD 3245284303;
#define pi acos(-1)
#define gcd __gcd
#define lcm(a, b) ((a) * (b) / gcd(a, b))
#define precision(n) cout << fixed << setprecision(n)
#define debug(arr) cerr << #arr << " = "; for (auto &x : arr) cerr << x << " "; cerr << endl;
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

struct Point3d {
	ld x, y, z; //Son vectores y si lo necesita en el plano, pues fije z=0
	Point3d() {}
	Point3d(ld x, ld y, ld z): x(x), y(y), z(z) {}
	Point3d& operator+=(const Point3d &t) {
		x += t.x;
		y += t.y;
		z += t.z;
		return *this;
	}
	Point3d& operator-=(const Point3d &t) {
		x -= t.x;
		y -= t.y;
		z -= t.z;
		return *this;
	}
	Point3d& operator*=(ld t) { //Aqui se opera a derecha!!
		x *= t;
		y *= t;
		z *= t;
		return *this;
	}
	Point3d& operator/=(ld t) { //Aqui se opera a derecha!!
		x /= t;
		y /= t;
		z /= t;
	return *this;
	}
	Point3d operator+(const Point3d &t) const {
		return Point3d(*this) += t;
	}
	Point3d operator-(const Point3d &t) const {
		return Point3d(*this) -= t;
	}
	Point3d operator*(ld t) const {
		return Point3d(*this) *= t;
	}
	Point3d operator/(ld t) const {
		return Point3d(*this) /= t;
	}
	static ld dot(Point3d a,Point3d b) {
		return a.x * b.x + a.y * b.y + a.z * b.z;
	}
	static Point3d cross(Point3d a, Point3d b) {
		return Point3d(a.y * b.z - a.z * b.y,
			a.z * b.x - a.x * b.z,
			a.x * b.y - a.y * b.x);
	}
	static ld norm(Point3d a) {//norma al cuadrado
		return dot(a,a);
	}
	static ld abs(Point3d a) {//norma normal xd
		return sqrt(norm(a));
	}
	static Point3d proj(Point3d a, Point3d b) {// proyeccion de a en b
		return b*(dot(a, b) / abs(b));
	}
	static ld angle(Point3d a, Point3d b) {
		return acos(dot(a, b) / abs(a) / abs(b));
	}
	bool operator==(const Point3d &other) const {
    		const double EPS = 1e-6;
		return fabs(x - other.x) < EPS && fabs(y - other.y) < EPS;
	}
};
Point3d operator*(ld a, Point3d b) {
	return b * a;
}




int main(){
	precision(10);
	ld t = 0.5;
	Point3d OA(3,1,4.5), OB(3,4,0);
	cout << "OA+OB = " << (OA+OB).x << " " <<(OA+OB).y << (OA+OB).z<< endl;
	cout << "OA-OB = " << (OA-OB).x << " " <<(OA-OB).y << (OA-OB).z<<endl;
	cout << "tOB = " << (OB*t).x << " " << (OB*t).y << (OB*t).z << endl;
	cout << "(1/t)OB = " << (OB/t).x << " " << (OB/t).y << (OB/t).z << endl;
	cout << "OAxOB = " << Point3d::cross(OA,OB).x << " " << Point3d::cross(OA,OB).y << " " << Point3d::cross(OA,OB).z << endl;
	cout << "OA*OB = " << Point3d::dot(OA,OB) << endl;
	cout << "||OB|| = " << Point3d::abs(OB) << endl;
	cout << "Proyeccion de OA sobre OB " << Point3d::proj(OA,OB).x << " " << Point3d::proj(OA,OB).y << " " << Point3d::proj(OA,OB).z << endl;
}